src/v/archival/tests/archiver_operations_impl_gtest.cc

```diff
@@ -0,0 +1,1275 @@
+        }();
+        auto delta = it->last - it->base + model::offset(1);
+        if (is_config) {
+            running_delta += delta;
+        } else {
+            // only add batch after the step interval (64KiB)
+            if (window >= sampling_step) {
+                auto ko = it->base - running_delta;
+                ix.add(it->base, ko, int64_t(it->physical_offset), it->ts_last);
+                window = 0;
+            }
+        }
+        window += it->size_bytes;
+
+        // Update stats
+        if (is_config) {
+            result.stats.total_conf_records += delta;
+        } else {
+            result.stats.total_data_records += delta;
+        }
+        if (result.stats.base_rp_offset == model::offset{}) {
+            result.stats.base_rp_offset = it->base;
+        }
+        result.stats.last_rp_offset = it->last;
+        if (result.stats.base_timestamp == model::timestamp{}) {
+            result.stats.base_timestamp = it->ts_base;
+        }
+        result.stats.last_timestamp = it->ts_last;
+        result.stats.size_bytes += it->size_bytes;
+    }
+
+    result.content = iobuf_to_bytes(ix.to_iobuf());
+    return result;
+}
+
+struct segment_desc {
+    model::offset base;
+    model::offset last;
+    model::offset_delta base_delta;
+    model::offset_delta last_delta;
+    model::timestamp base_ts;
+    model::timestamp last_ts;
+};
+
+auto make_upload(
+  const segment_desc& d, size_t upload_size, ss::input_stream<char> stream) {
+    auto prep_upl = std::make_unique<detail::prepared_segment_upload>();
+    prep_upl->is_compacted = false;
+    prep_upl->meta = segment_meta{
+      .is_compacted = false,
+      .size_bytes = upload_size,
+      .base_offset = d.base,
+      .committed_offset = d.last,
+      .base_timestamp = d.base_ts,
+      .max_timestamp = d.last_ts,
+      .delta_offset = d.base_delta,
+      .ntp_revision = expected_revision_id,
+      .archiver_term = expected_archiver_term,
+      .segment_term = expected_segment_term,
+      .delta_offset_end = d.last_delta,
+      .sname_format = segment_name_format::v3,
+    };
+    prep_upl->size_bytes = upload_size;
+    prep_upl->payload = std::move(stream);
+    prep_upl->offsets = inclusive_offset_range(d.base, d.last);
+    return prep_upl;
+}
+
+TEST_CORO(
+  archiver_operations_impl_test,
+  find_upload_candidates_success_1_segment_no_tx) {
+    scoped_config cfg;
+    cfg.get("storage_read_buffer_size").set_value(expected_read_buffer_size);
+    cfg.get("cloud_storage_segment_size_target")
+      .set_value(std::make_optional<size_t>(expected_target_size));
+    cfg.get("cloud_storage_segment_size_min")
+      .set_value(std::make_optional<size_t>(expected_min_size));
+
+    // Find single upload candidate
+    auto remote = ss::make_shared<remote_mock>();
+    auto pm = ss::make_shared<partition_manager_mock>();
+    auto builder = ss::make_shared<upload_builder_mock>();
+    auto partition = ss::make_shared<partition_mock>();
```

Comment:
nit: since these are repeated in most tests a fixture might be useful

Reply:
I meant more in terms of reducing repeated lines of code but it's not a big difference.